stem/model.py

```diff
@@ -191,6 +172,186 @@ def generate_straight_track(self, sleeper_distance: float, n_sleepers: int, rail
+                "coordinates": soil_equivalent_bottom,
+                "ndim": 0
+            }
+        }
+        self.gmsh_io.generate_geometry(constraint_model_soil_equivalent_part_settings, "")
+
+        constraint_model_soil_equivalent_part.get_geometry_from_geo_data(self.gmsh_io.geo_data,
+                                                                         constraint_model_soil_equivalent_name)
+
+        self.process_model_parts.append(constraint_model_soil_equivalent_part)
+
+    def get_points_outside_soil_volume(
+            self, model_part_name) -> Tuple[List[Union[int, Any]], List[Union[Sequence[float], Any]]]:
+        """
+        Get the points of the model part that are outside the volume of the model part.
+
+        Args:
+            - model_part_name (str): The name of the model part to check the points
+
+        Returns:
+            - List[int]: The ids of the points that are outside the volume of the model part.
+            - List[List[float]]: The coordinates of the points that are outside the volume of the model part.
+
+        """
+        # get bbox of the soil model parts
+        min_coords, max_coords = self.get_bounding_box_soil()
+
+        model_part = self.get_model_part_by_name(model_part_name)
+
+        if model_part is None:
+            raise ValueError(f"Model part {model_part_name} not found.")
+        else:
+            points_outside_volume = []
+            coordinates = []
+            if model_part.geometry is None:
+                raise ValueError(f"Model part {model_part_name} has no geometry.")
+            for point_id, point in model_part.geometry.points.items():
+                # dimensions except the out of plane direction
+                if self.ndim == 2:
+                    x_is_in = min_coords[0] <= point.coordinates[0] <= max_coords[0]
+                    y_is_in = min_coords[1] <= point.coordinates[1] <= max_coords[1]
+                elif self.ndim == 3:
+                    x_is_in = min_coords[0] <= point.coordinates[0] <= max_coords[0]
+                    y_is_in = min_coords[2] <= point.coordinates[2] <= max_coords[2]
+                # the z coordinate is the out of plane direction so it is not checked
+                if not x_is_in or not y_is_in:
+                    points_outside_volume.append(point_id)
+                    coordinates.append(point.coordinates)
+            return points_outside_volume, coordinates
+
+    def get_bounding_box_soil(self) -> Tuple[List[float], List[float]]:
+        """
+        Get the bounding box of the soil model parts.
+
+        Returns:
+            - Tuple[List[float], List[float]]: The minimum and maximum coordinates of the bounding box.
+        """
+        min_coords = [np.inf, np.inf, np.inf]
+        max_coords = [-np.inf, -np.inf, -np.inf]
+
+        for model_part in self.all_model_parts:
+            if isinstance(model_part, BodyModelPart) and isinstance(model_part.material, SoilMaterial):
```

Comment:
you can loop over body model parts instead of all model parts

Reply:
i understand, but you do for model_part in self.all_model_parts:, which can be for model_part in self.body_model_parts:

Reply:
as you dont want to also loop over conditions